File: backend/app/schemas/datasource.py
from pydantic import BaseModel
from datetime import datetime


class DataSourceBase(BaseModel):
    name: str
    type: str
    host: str
    port: int
    database: str = ""
    username: str = ""
    password: str = ""
    params: dict | None = None
    description: str | None = None


class DataSourceCreate(DataSourceBase):
    pass


class DataSourceUpdate(BaseModel):
    name: str | None = None
    type: str | None = None
    host: str | None = None
    port: int | None = None
    database: str | None = None
    username: str | None = None
    password: str | None = None
    params: dict | None = None
    description: str | None = None
    status: str | None = None


class DataSourceListItem(BaseModel):
    id: str
    name: str
    type: str
    host: str
    port: int
    database: str
    username: str
    password: str
    status: str
    table_count: int = 0
    enabled: bool = False
    description: str | None = None
    created_at: datetime
    updated_at: datetime
    model_config = {"from_attributes": True}


class DataSourceDetail(DataSourceBase):
    id: str
    status: str
    table_count: int = 0
    enabled: bool = False
    created_at: datetime
    updated_at: datetime
    created_by: str | None = None
    model_config = {"from_attributes": True}


class TestConnectionResult(BaseModel):
    success: bool
    message: str


class TableListResult(BaseModel):
    tables: list[str]


class TablePreviewResult(BaseModel):
    table: str
    columns: list[str]
    rows: list[list]

File: frontend/src/types/datasource.ts
export interface DataSource {
  id: string
  name: string
  type: string
  host: string
  port: number
  database: string
  username: string
  password: string
  params: Record<string, unknown> | null
  description: string | null
  status: string
  table_name: string
  record_count: number
  enabled: boolean
  created_at: string
  updated_at: string
  created_by: string | null
}

export interface DataSourceCreate {
  type: string
  host: string
  port: number
  database: string
  username: string
  password: string
  params?: Record<string, unknown> | null
  description?: string | null
}

export interface DataSourceUpdate {
  name?: string
  type?: string
  host?: string
  port?: number
  database?: string
  username?: string
  password?: string
  params?: Record<string, unknown> | null
  description?: string | null
  status?: string
}

export interface TestConnectionResult {
  success: boolean
  message: string
}
